Deprecate C Headers

CursedRock17 opened the pull request:
This pull request is an attempt to help #259, specifically the first task
> Rename all .h -> .hpp files. This will have to also have a deprecation warning, and should be done post-Foxy.

Since it's now post-Foxy I did something similar to this [`message_filters` pull request](https://github.com/ros2/message_filters/pull/135), which was to simply deprecate a `.h` file with a warning, include the `.hpp` file, and just copy paste the logic from the `.h` file into the `.hpp` file. There were some slight formatting changes that had to be made to make `cpplint` and `uncrustify` happy. There is currently a copyrighting problem within the [`LinearMath` files](https://github.com/ros2/geometry2/tree/rolling/tf2/include/tf2/LinearMath) since they seem to be from the `bullet` library. I would like to know where we should move from this point when it comes to those files.

CursedRock17 committed:
Deprecating C Headers

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

CursedRock17 committed:
Tunning header guards

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

CursedRock17 committed:
Fixing header order

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

ahcorde reviewed (changes requested):
copyright is failing

https://build.ros2.org/job/Rpr__geometry2__ubuntu_noble_amd64/123/testReport/junit/(root)/projectroot/copyright/

ahcorde commented on tf2/CMakeLists.txt:
```diff
@@ -55,13 +55,13 @@ if(BUILD_TESTING)
-    include/tf2/LinearMath/Scalar.h
-    include/tf2/LinearMath/Transform.h
-    include/tf2/LinearMath/Vector3.h
+    include/tf2/LinearMath/Matrix3x3.hpp
```
you should keep here both the `h` and the `hpp`

clalancette replied:
> you should keep here both the `h` and the `hpp`

Agreed.

clalancette commented:
Besides some of @ahcorde 's inline comments, we will also need to have PRs to the rest of the ROS 2 core to use the `.hpp` files instead of the `.h` ones.  Also, this is a large enough change that once we are ready to merge it, we should probably send a message to https://discourse.ros.org letting people know that we've done this.

CursedRock17 committed:
Fixing Copyrighting

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

CursedRock17 commented:
Yeah I figured we would have to extend PRs to various libraries, I have one linked from `rviz`, but I haven't found anymore, I'll have to do some digging around to find out if I've missed any more of the core ones. As far as the copyrighting goes, I believe those inline comments were the missing piece, I've submitted a fix.

CursedRock17 commented:
Updated downstream packages: 
  - https://github.com/ros2/rviz/pull/1289
  - https://github.com/ros2/ros2_documentation/pull/4828
  - https://github.com/ros2/cartographer_ros/pull/75
  - https://github.com/ros/geometry/pull/244
  - https://github.com/ros/geometry2/pull/573
  - https://github.com/ros/geometry_tutorials/pull/87
  - https://github.com/ros-perception/imu_pipeline/pull/25
  - https://github.com/ros-perception/laser_geometry/pull/98
  - https://github.com/ros-perception/laser_filters/pull/206
  - https://github.com/ros-perception/image_pipeline/pull/1039
  - https://github.com/ros-perception/perception_pcl/pull/469
  - https://github.com/ros-controls/ros2_controllers/pull/1325
  - https://github.com/ros-navigation/navigation2/pull/4732
  - https://github.com/ros-industrial/industrial_training/pull/420

These were all the repositories that I could find, within the core ROS organizations, in which `tf2` was utilized. There are likely thousands of repositories which are going to have warnings, which are warnings as to not break API, but I certainly wouldn't mind making a general PSA.

Timple commented:
Would it make sense to provide `.hpp` files for all active distros? With as only content the include of the `.h` file. 

This would mean all projects that do not have a branch per release can use this syntax already. Preventing ugly statements like this everywhere: https://github.com/autowarefoundation/autoware.universe/blob/9bd0f77c255edcff129ff08d190a5fdd8c45b6dc/localization/yabloc/yabloc_common/src/pub_sub.cpp#L18

For rolling we can keep the deprecation flag, not for older distros.

CursedRock17 commented:
Yeah I don't see much of a compile downside, just a weirder file structure for older distributions. I assume we would just never phase out the `.h` files unlike the rolling branch in which we can remove them in `L-Turtle`. I would be able to add this "backport" later today if that's something we agree on.

Also there already seems to be some consensus in the other core repositories:
 - https://github.com/ros-perception/perception_pcl/pull/469
>  we should create backwards compatibilty headers that emit compile warnings so users have time to migrate to the new headers.
 - https://github.com/ros-controls/ros2_controllers/pull/1325
> Can you add conditioning in the code so that from rolling on it uses the .hpp, and in the prior versions it uses the .h version?

CursedRock17 commented:
The following pull requests are the "backports" that will prevent gross include macros with the formulating of `.hpp` files for ROS' non-EOL distributions. Do I need to go back any further, or will these be alright for now?
 - #722 
 - #723
 - #724

Timple commented:
Iron I guess?

clalancette reviewed (changes requested):
There is one major change I'd like to see in this PR, and that is to *not* make the linting updates to the tf2/include/tf2/LinearMath headers.  While I think we may eventually want to do this, for now I think it will be easier to track provenance without doing that.

Once that is fixed, and we run CI on this and the downstream PRs, I think we can get this in.

CursedRock17 committed:
Removing LinearMath C++ Header

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

CursedRock17 commented:
Changes have been implemented, that being said was it not just easier to do this now opposed to later, especially when naming conventions are going to be mixed?
Also I'll begin making necessary changes downstream.

clalancette commented:
> Changes have been implemented, that being said was it not just easier to do this now opposed to later, especially when naming conventions are going to be mixed?
> Also I'll begin making necessary changes downstream.

No, sorry, that's not what I meant.  I meant leave the `.hpp` files, but do not do things like remove the trailing whitespace.

CursedRock17 commented:
Ohhh, I was quite confused lol, I will be reverting the changes in addition to the whitespace changes.

CursedRock17 committed:
Revert "Removing LinearMath C++ Header"

This reverts commit 62879df5ada52efb3a7efded21a45ee14c5f718d.

Ryanf55 commented:
> Iron I guess?

Can you do humble if it's not too much work?

CursedRock17 committed:
Linter Matrix3x3.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Matrix3x3.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter MinMax.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter MinMax.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter QuadWord.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter QuadWord.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Quaternion.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Quaternion.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Scalar.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Scalar.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Transform.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Transform.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Vector3.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Linter Vector3.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

clalancette commented on tf2/include/tf2/visibility_control.h:
This particular file is actually C code, and mostly internal, so I think we do not need to rename this.  That is, keep *just* this file as `visibility_control.h`.

clalancette reviewed (changes requested):
All right, one more relatively small thing to fix, then I think this will be ready for CI.

CursedRock17 committed:
Reverting visibility_control changes

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

Visibility Controls utils.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Visibility Controls convert.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Visibility Controls  Quaternion.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Visibility Controls Vector3.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Visibility Controls  QuadWord.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Visibility Controls Transform.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Visibility Controls  exceptions.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Visibility Controls time_cache.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Visibility Controls transform_storage.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update time.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update buffer_core.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update buffer_core_interface.hpp

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

CursedRock17 commented:
When I get back to my workstation later today, I can roll these improvements (whitespacing & visibility_control) onto the other branches I adjusted for the backport.

clalancette commented:
Pulls: https://github.com/ros2/geometry2/pull/720, https://github.com/ros2/rviz/pull/1289, https://github.com/ros-perception/laser_geometry/pull/98
Gist: https://gist.githubusercontent.com/clalancette/b83127091186f2489c09c4c787585861/raw/3c16fd56b4d645d09b9cadf800da79cbfd1acbaf/ros2.repos
BUILD args: 
TEST args: 
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14792
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22096)](http://ci.ros2.org/job/ci_linux/22096/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16353)](http://ci.ros2.org/job/ci_linux-aarch64/16353/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1616)](http://ci.ros2.org/job/ci_linux-rhel/1616/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22837)](http://ci.ros2.org/job/ci_windows/22837/)

clalancette commented:
So there are at least two problems here:

1.  We need a PR to `interactive_markers` to deal with the change.
2.  Windows doesn't seem to like the `#warning` construct, so we may have to do something special for Windows there.  Maybe something like https://github.com/ros2/common_interfaces/blob/473511f5807610bb8ed7744dfd2986dba4601fa6/sensor_msgs/include/sensor_msgs/point_cloud_conversion.hpp#L41-L49

CursedRock17 committed:
Fixing

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

Update Matrix3x3.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update MinMax.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update QuadWord.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update Quaternion.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update Scalar.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update Transform.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update Vector3.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update buffer_core.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update buffer_core_interface.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update convert.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update exceptions.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update convert.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update utils.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update time.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update time_cache.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update transform_datatypes.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update transform_storage.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update utils.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update MinMax.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update QuadWord.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update Quaternion.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update Scalar.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update Transform.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update Vector3.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update buffer_core.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update buffer_core_interface.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update convert.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update exceptions.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update convert.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update utils.h

Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

CursedRock17 committed:
Whitespacing Issues

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

CursedRock17 commented:
Hopefully windows likes the current implemention. Also backport PRs are ready to go as well.

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22106)](http://ci.ros2.org/job/ci_linux/22106/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16361)](http://ci.ros2.org/job/ci_linux-aarch64/16361/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1625)](http://ci.ros2.org/job/ci_linux-rhel/1625/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22848)](http://ci.ros2.org/job/ci_windows/22848/)

ahcorde reviewed (changes requested):
There are still some warnings

CursedRock17 commented:
I think it's just the repos being built, I changed up `interactive_markers` which seems to be the repository throwing out some [warnings](https://ci.ros2.org/job/ci_windows/22848/msbuild/folder.825285168/#issuesContent) as it's still on the `rolling` branch. My updated branch stops using the `.h` files that are the only ones triggering the warnings, which should be ok.

MichaelOrlov commented:
@ahcorde Could you please try to rerun CI?

ahcorde commented:
included PRs: geometry2, laser_geometry, RViz2 and interactive_markers

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22284)](http://ci.ros2.org/job/ci_linux/22284/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16544)](http://ci.ros2.org/job/ci_linux-aarch64/16544/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1811)](http://ci.ros2.org/job/ci_linux-rhel/1811/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23073)](http://ci.ros2.org/job/ci_windows/23073/)

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22286)](http://ci.ros2.org/job/ci_linux/22286/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16546)](http://ci.ros2.org/job/ci_linux-aarch64/16546/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1813)](http://ci.ros2.org/job/ci_linux-rhel/1813/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23075)](http://ci.ros2.org/job/ci_windows/23075/)

Ryanf55 commented:
> > Iron I guess?
> 
> Can you do humble if it's not too much work?

What's the story on humble? Our work environment is now broken from the docker images supplied from `ros:humble-ros-core`. 

Currently, in this broken state, we have the following installed.
* ros-humble-tf2 version 0.25.9-1jammy.20241128.023902
* ros-humble-tf2-eigen version 0.25.11-1jammy.20250114.170617

In tf2_eigen.hpp, it includes `"tf2/convert.hpp"`.

Then, the files that exist:
```
$ ls /opt/ros/humble/include/tf2/tf2/ -1
LinearMath
buffer_core.h
buffer_core_interface.h
convert.h
exceptions.h
impl
time.h
time_cache.h
transform_datatypes.h
transform_storage.h
utils.h
visibility_control.h
```

Where's the deprecation period and how did this get all the way through the build farm into a docker image with debians that don't work together. Why isn't `tf2-eigen` expressing a minimum version on `tf2` so it fails at configure time?

Sorry to be blunt, but this is just frustrating.

clalancette commented:
> * ros-humble-tf2 version 0.25.9-1jammy.20241128.023902
> 
> * ros-humble-tf2-eigen version 0.25.11-1jammy.20250114.170617

These aren't the same versions, which is why you are having trouble here.  The debians that are released should all match as 0.25.11; the fact that they aren't probably means your local scripts are missing an apt-get update somewhere (i.e. it is using a cached layer with tf2, and a new layer with tf2_eigen).

Ryanf55 commented:
Thanks. If there's any changes we could do to ensure that we fail at configure time rather than build time, I'd be happy to contribute those.
